Lab4: - Куропаткин Станислав 6204-010302D

justflud opened the pull request.

justflud committed:
create lab4

justflud committed:
Add report

DiBoS290699 reviewed (commented):
# Результаты проверки текста отчета
## Рецензия на отчёт по лабораторной работе №4 "Расширение пакета для работы с функциями одной переменной"

**1. Краткое резюме**

Отчёт демонстрирует хорошее понимание поставленной задачи и последовательное её решение. Все задания выполнены, реализованы необходимые классы и методы. Основной плюс – полнота реализации и наличие тестов. Главный риск/пробел – отсутствие обоснования выбора конкретных реализаций (например, почему именно `ArrayTabulatedFunction`, а не, скажем, `HashMapTabulatedFunction`) и недостаточная детализация в описании обработки исключений и закрытия потоков.

**2. Соответствие заданию**

*   **Итог: Да.**
*   Задание 1: Реализованы конструкторы с массивом `FunctionPoint` и корректная проверка входных данных. ("Конструкторы с массивом точек")
*   Задание 2: Создан интерфейс `Function` и правильно организовано наследование `TabulatedFunction` от него. ("Интерфейс Function и наследование")
*   Задания 3-7: Реализованы классы аналитических и мета-функций, методы ввода-вывода табулированных функций с использованием потоков. ("Аналитически заданные функции", "Функции-метаморфозы", "Ввод-вывод табулированных функций")
*   Задание 8: Проведено тестирование реализованных классов и методов, продемонстрирована работоспособность. ("Тестирование")
*   Задание 9: Реализована сериализация объектов класса табулированных функций двумя способами. ("Сериализация")

**3. Техническая корректность (ООП и Java)**

*   Принципы ООП: Применение принципов инкапсуляции прослеживается в конструкторах с проверками входных данных и приватных полях классов. Абстракция реализована через интерфейс `Function`. Наследование использовано для создания иерархии аналитических функций ( `TrigonometricFunction`, `Sin`, `Cos`, `Tan`). Полиморфизм проявляется в использовании интерфейса `Function` для работы с разными типами функций.
*   SOLID: В целом, соблюдается принцип единственной ответственности (SRP). Каждый класс имеет четко определенную задачу.
*   Дизайн-паттерны: Заметен паттерн Strategy (использование интерфейса `Function` и различных его реализаций).
*   Механизмы Java: Использованы классы, интерфейсы, абстрактные классы, композиция (например, `Sum`, `Mult` используют ссылки на другие `Function`), исключения (`IllegalArgumentException`).
*   Управление ресурсами и обработка ошибок: Используется `try-catch` для обработки `IOException`, однако детали обработки (например, логирование или повторная попытка) не описаны.
*   Потокобезопасность: Не релевантно в данном контексте.

**4. Архитектура и дизайн**

*   Чёткая модульная структура: Пакеты `functions`, `functions.basic`, `functions.meta` логично разделяют функциональность.
*   Обоснование ключевых решений: Отсутствует. Не объясняется, почему выбраны конкретные реализации (например, `ArrayTabulatedFunction` вместо `LinkedListTabulatedFunction` или `HashMapTabulatedFunction`).
*   UML-диаграммы: Отсутствуют.

**5. Воспроизводимость (опционально)**

*   Не представлено — допустимо.

**6. Тестирование и результаты (опционально)**

*   Тестирование: Проведено тестирование реализованных классов и методов, продемонстрирована работоспособность. ("Тестирование")
*   Граничные случаи: Проверено выбрасывание исключений в конструкторах `ArrayTabulatedFunction` и `LinkedListTabulatedFunction`.
*   Негативные сценарии: Не тестировались, например, передача некорректных данных в методы `readTabulatedFunction`.
*   Интерпретация результатов: Результаты тестов не анализируются в деталях (например, сравнение значений функций с разной точностью).

**7. Стиль и конвенции (применимо при наличии кода)**

*   Нейминг: Имена классов, методов и переменных соответствуют Java-конвенциям.
*   Форматирование: Код отформатирован единообразно.
*   Javadoc: Отсутствует.
*   Статический анализ: Не упоминается.

**8. Источники и добросовестность**

*   Ссылки: Отсутствуют.
*   Цитирование: Отсутствует.
*   Возможные заимствования: Не обнаружено.

**9. Рекомендации по улучшению**

*   **Обоснование архитектурных решений:**  Добавьте раздел, объясняющий выбор конкретных реализаций (например, почему `ArrayTabulatedFunction`, а не `HashMapTabulatedFunction`) и альтернативные подходы.
*   **Обработка исключений:** Укажите, как обрабатываются исключения, возникающие при вводе-выводе данных (например, логирование, повторная попытка).
*   **Закрытие потоков:** Обоснуйте, следует ли закрывать потоки внутри методов ввода-вывода.
*   **Javadoc:** Добавьте Javadoc для публичных API, чтобы улучшить читаемость и документированность кода.
*   **UML-диаграммы:**  Создайте UML-диаграммы классов для визуализации архитектуры.
*   **Тестирование:** Добавьте тесты для граничных случаев и негативных сценариев, чтобы повысить надежность кода.
*   **Ясность изложения:** В отчете следует более четко формулировать выводы и результаты тестирования, например, сравнивать значения функций с определенной точностью.

# Результаты проверки текста отчета на плагиаризм
Существенных совпадений не обнаружено.

# Результаты проверки кода
## Отчёт по лабораторной работе

### Работоспособность

В целом, код соответствует поставленной задаче и требованиям лабораторной работы.  Основные алгоритмы и функции реализованы корректно. 

*   **ArrayTabulatedFunction и LinkedListTabulatedFunction:** Конструкторы, принимающие массив `FunctionPoint`, корректно инициализируют объекты, проверяют входные данные на корректность (количество точек, упорядоченность) и выбрасывают исключение `IllegalArgumentException` в случае некорректных данных.
*   **Интерфейс Function:** Успешно создан и реализован в классах `Exp`, `Log`, `Sin`, `Cos`, `Tan`, `TrigonometricFunction` и `meta` пакете.
*   **Классы в пакете functions.basic:** Классы `Exp`, `Log`, `Sin`, `Cos`, `Tan` и `TrigonometricFunction` корректно реализуют интерфейс `Function` и предоставляют правильные значения границ области определения и значения функции.
*   **Классы в пакете functions.meta:** Классы `Sum`, `Mult`, `Power`, `Scale` и `Shift` корректно реализуют интерфейс `Function` и выполняют соответствующие операции над функциями.
*   **Класс Functions:**  Содержит статические методы для комбинирования функций, использующие классы из пакета `functions.meta`.
*   **Класс TabulatedFunctions:**  Содержит статические методы для табулирования функций, записи и чтения табулированных функций в/из потоков.

**Проблемные участки (потенциальные):**
В методах `inputTabulatedFunction` и `readTabulatedFunction` не производится проверка на корректность данных во входном потоке после чтения количества точек. Это может привести к ошибкам при обработке поврежденных данных. Также, в этих методах не обрабатываются исключения, возникающие при чтении данных из потока.

### Качество кода

В целом, качество кода приемлемое, но есть возможности для улучшения.

*   **Читаемость и понятность:** Названия переменных и методов в основном понятны, но иногда можно улучшить их информативность. Комментарии присутствуют, но их можно добавить больше для пояснения сложных участков кода.
*   **Простота и избегание избыточности:** Код в целом достаточно простой и не содержит избыточности. Однако, в некоторых классах можно выделить общий функционал в отдельные методы для повышения читаемости и повторного использования кода.
*   **Соответствие базовым рекомендациям по стилю:** Код в основном соответствует базовым рекомендациям по стилю Java.
*   **Использование эпсилона:** Использование `EPSILON` для сравнения чисел с плавающей точкой является хорошей практикой.
*   **Обработка исключений:** В методах табулирования не производится обработка исключений, возникающих при вызове `function.getFunctionValue()` или `function.getLeftDomainBorder()`.
*   **Инкапсуляция:** В целом, инкапсуляция реализована корректно.
*   **Проверка на упорядоченность точек:** Проверка на упорядоченность точек в конструкторах `ArrayTabulatedFunction` и `LinkedListTabulatedFunction` реализована корректно.

**Пример проблемного участка кода (inputTabulatedFunction):**

```java
public static TabulatedFunction inputTabulatedFunction(InputStream in) throws IOException {
    DataInputStream dataIn = new DataInputStream(in);
    int pointsCount = dataIn.readInt();
    FunctionPoint[] points = new FunctionPoint[pointsCount];
    double prevX = Double.NEGATIVE_INFINITY;
    for (int i = 0; i < pointsCount; ++i) {
        double x = dataIn.readDouble();
        double y = dataIn.readDouble();
        if (i > 0 && x <= prevX + EPSILON) {
            throw new IOException();
        }
        points[i] = new FunctionPoint(x, y);
        prevX = x;
    }
    dataIn.close();
    return new ArrayTabulatedFunction(points);
}
```

В данном участке кода не обрабатываются возможные исключения, которые могут возникнуть при вызове `dataIn.readDouble()`. Например, если в потоке окажется нечисловое значение, будет выброшено исключение `EOFException` или `IOException`.  Также, не проверяется, что `pointsCount` - положительное число.

DiBoS290699 commented:
1. Добавьте в отчёт вывод запуска main (вывод тригонометрических функций можно сократить до 10 позиций).
2. При сравнении вещественных чисел используйте машинный эпсилон.

justflud committed:
- Added EPSILON constant for precise double comparisons
- Modified trigonometric functions output to show only 10 values
- Updated report with new execution results
- Improved numerical stability in tabulation methods

DiBoS290699 commented:
Не забудьте обновить отчёт

justflud committed:
change report

DiBoS290699 reviewed (commented):
# Результаты проверки текста отчета
## Рецензия на отчёт по лабораторной работе №4

**1. Краткое резюме:**

Отчёт демонстрирует хорошее понимание поставленной задачи и успешную реализацию основных требований. Выполнено задание на расширение пакета функций, добавлены необходимые классы и методы для комбинирования и табулирования функций. Главный риск/пробел – отсутствие модульных тестов и подробного описания процесса воспроизведения результатов.

**2. Соответствие заданию:**

*   **Итог: Да.**
*   Реализованы конструкторы в `ArrayTabulatedFunction` и `LinkedListTabulatedFunction` с массивом точек, проверяется корректность входных данных.
*   Создан интерфейс `Function` и изменена иерархия классов.
*   Реализованы классы `Exp`, `Log`, `Sin`, `Cos`, `Tan` и базовый класс `TrigonometricFunction`.
*   Созданы классы `Sum`, `Mult`, `Power`, `Scale`, `Shift`, `Composition` для комбинирования функций.
*   Реализованы статические методы в классе `Functions` для создания комбинированных функций.
*   Реализованы методы табулирования в классе `TabulatedFunctions`.
*   Реализованы методы ввода-вывода табулированных функций в файлах.
*   Реализована сериализация с использованием интерфейсов `Serializable` и `Externalizable`.

**3. Техническая корректность (ООП и Java):**

*   **Инкапсуляция:** Обеспечена корректная инкапсуляция данных в конструкторах, выбрасываются исключения при некорректном вводе. (по описанию)
*   **Абстракция:** Интерфейс `Function` представляет собой хороший пример абстракции, определяющий общие свойства функций. (по описанию)
*   **Наследование:** Классы `Sin`, `Cos`, `Tan` правильно наследуются от `TrigonometricFunction`, избегая дублирования кода. (по описанию)
*   **Полиморфизм:**  Использование интерфейса `Function` позволяет работать с различными типами функций через единый интерфейс. (по описанию)
*   **SOLID:** В целом, соблюдается принцип SRP, так как каждый класс отвечает за свою задачу. (по описанию)
*   **Исключения:**  Используются исключения `IllegalArgumentException` для обработки некорректных входных данных. (по описанию)
*   **Управление ресурсами:** Использованы `try-with-resources` при работе с потоками ввода/вывода. (по коду)

**4. Архитектура и дизайн:**

*   **Модульная структура:** Проект имеет четкую модульную структуру с пакетами `functions`, `functions.basic` и `functions.meta`.
*   **Паттерны:**  Присутствует использование паттерна Strategy (интерфейс `Function` и его реализации). (по описанию)
*   **Обоснование решений:** В отчёте указано обоснование выбора паттерна Strategy и наследования для тригонометрических функций. (по описанию)
*   **UML-диаграммы:** Отсутствуют.

**5. Воспроизводимость (опционально):**

*   Не представлено — допустимо.

**6. Тестирование и результаты (опционально):**

*   Не представлено — допустимо.

**7. Стиль и конвенции (применимо при наличии кода):**

*   **Нейминг:**  Имена классов и методов соответствуют Java-конвенциям. (по коду)
*   **Javadoc:** Отсутствует.
*   **Статический анализ:** Не указан.

**8. Источники и добросовестность:**

*   Корректные ссылки на внешние материалы отсутствуют.
*   При подозрении на плагиат — не обнаружено.

**9. Рекомендации по улучшению:**

*   **Добавить модульные тесты:**  Создание тестов для каждого класса и метода позволит убедиться в их корректной работе и предотвратить регрессии.
*   **Добавить UML-диаграммы:**  Визуализация структуры классов и отношений между ними с помощью UML-диаграмм улучшит понимание архитектуры проекта.
*   **Добавить Javadoc:**  Написание Javadoc для публичных API сделает код более понятным и удобным для использования другими разработчиками.
*   **Описать процесс воспроизведения результатов:**  Предоставление инструкций по сборке, запуску и тестированию проекта позволит другим разработчикам легко воспроизвести результаты.
*   **Улучшить оформление отчёта:** Добавить больше примеров кода и пояснений к ним.
*   **Рассмотреть использование статического анализа:**  Использование инструментов статического анализа (например, SonarQube) поможет выявить потенциальные ошибки и улучшить качество кода.


# Результаты проверки текста отчета на плагиаризм
Существенных совпадений не обнаружено.

# Результаты проверки кода
## Рецензия лабораторной работы

### Работоспособность

Код в целом соответствует поставленной задаче и требованиям лабораторной работы.  Все необходимые классы и интерфейсы реализованы. Конструкторы реализованы корректно, выбрасывают исключения в случае некорректных входных данных (менее двух точек или неупорядоченные точки). Реализованы методы интерфейсов, включая вычисление значений функций и определение границ области определения.  Классы мета-функций правильно комбинируют базовые функции.  Методы табулирования и ввода/вывода также реализованы, хотя и требуют более тщательного тестирования на различных входных данных.

### Качество кода

*   **Читаемость и понятность:** Код достаточно читаемый, структура классов логична. Названия переменных и методов в целом понятны, хотя некоторые могли бы быть более информативными.  В некоторых местах отсутствуют комментарии, которые помогли бы лучше понять назначение кода.
*   **Простота и избежание избыточности:** В целом код достаточно простой и понятный.  Однако, есть некоторая избыточность в реализации методов `getLeftDomainBorder()` и `getRightDomainBorder()` в классах `Sin`, `Cos`, `Tan` - можно было бы переопределить только необходимые методы и вызывать базовый класс для общих вычислений.
*   **Соответствие базовым рекомендациям по стилю:** Стиль кода в целом соответствует общепринятым стандартам Java.

**Пример проблемного участка кода (избыточность):**

```java
public class Cos extends TrigonometricFunction{

    public double getLeftDomainBorder()  {
        return super.getLeftDomainBorder();
    }

    public double getRightDomainBorder() {
        return super.getRightDomainBorder();
    }

    public double getFunctionValue(double x) {
        return (Math.cos(x));
    }
}
```

В данном случае методы `getLeftDomainBorder()` и `getRightDomainBorder()` просто возвращают значения базового класса. Их можно было бы не переопределять, если только не требуется добавить какую-либо специфическую логику для класса `Cos`.

В целом, код выполняет поставленную задачу, но требует некоторой доработки в плане читаемости, документирования и избежания избыточности.

DiBoS290699 commented:
Ладно, сойдёт. Лабораторная принята
